### PR TITLE
Dog status codes 

### LINKED ISSUE #568: Dog status codes
## Description
A command such as `.http dogs` would be implemented to visually show what different status codes are, with dogs.

## Reasoning
Another fun way to learn status codes or refer them. And for non-cat people and dog lover ❤️ .

## Proposed Implementation
- Make a group named `http` which would contain `cat` and `dog` nested commands.
- Since dogs have some codes missing, we will use `aiohttp` and `regex` to get the status codes using `commands.task` which will run every 2 hours probably.
- This would implemented in a similar way like `.http_cat`.

## Would you like to implement this yourself?

- [x] I'd like to implement this feature myself
- [ ] Anyone can implement this feature

### PR DESCRIPTION
**Pull request by Shivansh-007**
## Relevant Issues

Closes #568 

## Description
HTTP status dogs doesn't the following codes: `[101, 102, 421, 511, 499, 599, 415]`.
I have made a group for `cat` and `dog` status codes and dog status code is just duplicate to that of the cat just it adds another check by checking the response status after requesting the url.

## Screenshots
Screenshot from 2021-01-31 05-19-21
Screenshot from 2021-01-31 05-19-39

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] **Lint your code** (`pipenv run lint`)?
- [x] Set the PR to **allow edits from contributors**?

### CODE REVIEW

File: bot/exts/evergreen/status_codes.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        try:
+            HTTPStatus(code)
+            async with self.bot.http_session.get(
+                    f'https://httpstatusdogs.com/img/{code}.jpg',
+                    allow_redirects=False
+            ) as response:
+                if response.status != 302:
+                    embed.set_image(url=f'https://httpstatusdogs.com/img/{code}.jpg')
+                else:
+                    raise ValueError
```
Comment (ChrisLovering) [inline bot/exts/evergreen/status_codes.py:50]: Could do: raise `NotImplementedError` instead, and handle it separately from `ValueError`. This way we inform user if the code doesn't exist at all, or if its just not been implemented by httpstatusdogs.com

File: bot/exts/evergreen/status_codes.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+from http import HTTPStatus
+
+import discord
+from discord.ext import commands
+
+
+class StatusCats(commands.Cog):
```
Comment (ChrisLovering) [inline bot/exts/evergreen/status_codes.py:7]: Should do: Rename the cog, now that it does more than just cats.

File: bot/exts/evergreen/status_codes.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        embed = discord.Embed(title=f'**Status: {code}**')
+
+        try:
+            HTTPStatus(code)
+
+        except ValueError:
+            embed.set_footer(text='Inputted status code does not exist.')
+
+        else:
+            embed.set_image(url=f'https://http.cat/{code}.jpg')
```
Comment (ChrisLovering) [inline bot/exts/evergreen/status_codes.py:31]: Could do: `http.cat` raises a 404 on some status codes that `HTTPStatus` says is a valid code, so we could handle those in a similar way to what you have done for the dog command.

**Review by ChrisLovering — CHANGES_REQUESTED** [review]
I think we should rename the cog. The two could do's are optional, if you feel like implementing them. If not, they can be raised as separate issues :)

File: bot/exts/evergreen/status_codes.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        try:
+            HTTPStatus(code)
+            async with self.bot.http_session.get(
+                    f'https://http.cat/{code}.jpg',
+                    allow_redirects=False
+            ) as response:
+                if response.status != 404:
+                    embed.set_image(url=f'https://http.cat/{code}.jpg')
+                else:
+                    raise NotImplementedError
```
Comment (ChrisLovering) [inline bot/exts/evergreen/status_codes.py:33]: Must do: Add an except block for this raised error, like you have for dogs

**Review by ChrisLovering — CHANGES_REQUESTED** [review]
Just 1 change outstanding. I've tested the rest of this locally and it seems good to me! Nice job :)

**Comment by Xithrius** [comment]
When you input 101 for the dog status codes, what does it output? A 404?

**Comment by ChrisLovering** [comment]
> When you input 101 for the dog status codes, what does it output? A 404?

image

**Comment by Xithrius** [comment]
Sorry, what I mean is when requesting from the url for a 101, what does it return?

**Comment by ChrisLovering** [comment]
> Sorry, what I mean is when requesting from the url for a 101, what does it return?

My bad, any code that hasn't been implemented redirects to `https://httpstatusdogs.com/404-not-found` which is a page with a 404 image, but a bunch of other content. This is the case for 101, but also 999 as example.

**Comment by Xithrius** [comment]
I see, thank you for this information.

**Comment by Shivansh-007** [comment]
@ChrisLovering Mentioned the requested changes!

**Review by ChrisLovering — APPROVED** [review]
LGTM!

File: bot/exts/evergreen/status_codes.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+
+    @commands.group(name="http_status", aliases=("status", "httpstatus"))
+    async def http_status_group(self, ctx: commands.Context) -> None:
+        """Group containing dog and cat http status code commands."""
+        if not ctx.invoked_subcommand:
+            await ctx.send_help(ctx.command)
+
+    @http_status_group.command(aliases=['cat'])
```
Comment (Xithrius) [inline bot/exts/evergreen/status_codes.py:19]:
```suggestion
    @http_status_group.command(name='cat')
```

File: bot/exts/evergreen/status_codes.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        except ValueError:
+            embed.set_footer(text='Inputted status code does not exist.')
+
+        except NotImplementedError:
+            embed.set_footer(text='Inputted status code is not implemented by http.cat yet.')
+
+        finally:
+            await ctx.send(embed=embed)
+
+    @http_status_group.command(aliases=['dog'])
```
Comment (Xithrius) [inline bot/exts/evergreen/status_codes.py:44]:
```suggestion
    @http_status_group.command(name='dog')
```

File: bot/exts/evergreen/status_codes.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    async def http_dog(self, ctx: commands.Context, code: int) -> None:
+        """Sends an embed with an image of a dog, portraying the status code."""
+        embed = discord.Embed(title=f'**Status: {code}**')
+
+        try:
+            HTTPStatus(code)
+            async with self.bot.http_session.get(
+                    f'https://httpstatusdogs.com/img/{code}.jpg',
```
Comment (Xithrius) [inline bot/exts/evergreen/status_codes.py:52]: URL should be a constant.

File: bot/exts/evergreen/status_codes.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                    raise NotImplementedError
+
+        except ValueError:
+            embed.set_footer(text='Inputted status code does not exist.')
+
+        except NotImplementedError:
+            embed.set_footer(text='Inputted status code is not implemented by httpstatusdogs.com yet.')
+
+        finally:
+            await ctx.send(embed=embed)
```
Comment (Xithrius) [inline bot/exts/evergreen/status_codes.py:67]: Good use of the `finally` statement.

File: bot/exts/evergreen/status_codes.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        embed = discord.Embed(title=f'**Status: {code}**')
+
+        try:
+            HTTPStatus(code)
+            async with self.bot.http_session.get(
+                    f'https://httpstatusdogs.com/img/{code}.jpg',
+                    allow_redirects=False
+            ) as response:
+                if response.status != 302:
+                    embed.set_image(url=f'https://httpstatusdogs.com/img/{code}.jpg')
```
Comment (Xithrius) [inline bot/exts/evergreen/status_codes.py:56]: Once this URL is a constant, you'll only have to only format this once at the start of this function.

**Comment by Shivansh-007** [comment]
Mentioned the requested changes @Xithrius !

**Review by Xithrius — APPROVED** [review]

**Pull request merged**